protocols/kad/src/behaviour.rs

```diff
@@ -2704,7 +2714,10 @@ pub type GetProvidersResult = Result<GetProvidersOk, GetProvidersError>;
 #[derive(Debug, Clone)]
 pub struct GetProvidersOk {
     pub key: record::Key,
+    /// The new set of providers discovered.
     pub providers: HashSet<PeerId>,
+    /// How many providers have been discovered so var.
+    pub providers_so_far: usize,
     pub closest_peers: Vec<PeerId>,
```

Comment:
Is `closest_peers` helpful on any of the `OutboundQueryProgressed` but the very last one? Maybe we can split `GetProvidersOk` via an enum similar to `GetRecordOk`:

``` rust 
enum GetProvidersOk {
  FoundProviders {
    ...
  },
  FinishedWithNoAdditionalRecord {
    closest_peers: Vec<PeerId>,
  }
}
```

If I am not mistaken we currently also emit a `OutboundQueryProgressed` with an empty `providers` in `query_finished`. With the above, this would no longer need to be the case.

Reply:
yes, very good idea, done